service/scaladsl/client/src/main/scala/com/lightbend/lagom/scaladsl/client/ServiceClient.scala

```diff
@@ -140,28 +150,187 @@ trait LagomServiceClientComponents extends TopicFactoryProvider { self: LagomCon
+ * and a [[ConfigurationServiceLocator]]. The easiest way to use these is to mix in their respective Components traits.
+ *
+ * For example, here’s a client application built using the static service locator, which uses a static URI:
+ * {{{
+ * import java.net.URI
+ * import com.lightbend.lagom.scaladsl.client._
+ * import play.api.libs.ws.ahc.AhcWSComponents
+ *
+ * val clientApplication = new StandaloneLagomClientFactory("my-client")
+ *   with StaticServiceLocatorComponents
+ *   with AhcWSComponents {
+ *
+ *   override def staticServiceUri = URI.create("http://localhost:8080")
+ * }
+ * }}}
+ *
+ *
+ * @param clientName The name of the service that is consuming the Lagom service. This will impact how calls made through clients
+ *                   generated by this factory will identify themselves.
+ * @param classLoader A classloader, it will be used to create the service proxy and needs to have the API for the client in it.
+ */
+abstract class StandaloneLagomClientFactory(
+  clientName:  String,
+  classLoader: ClassLoader = classOf[StandaloneLagomClientFactory].getClassLoader
+) extends LagomClientFactory(clientName, classLoader) {
+
+  lazy val actorSystem: ActorSystem = new ActorSystemProvider(environment, configuration, applicationLifecycle).get
+  override lazy val materializer: Materializer = ActorMaterializer.create(actorSystem)
+
+  private val log = LoggerFactory.getLogger(this.getClass)
+
+  /**
+   * Stop this [[LagomClientFactory]] by shutting down the internal [[akka.actor.ActorSystem]], Akka Streams [[akka.stream.Materializer]]
+   * and internal resources.
+   */
+  override def stop(): Unit = {
+
+    // we need to use the stop method from ApplicationLifecycle because the
+    // WebSocket client register a stop hook on it
+    // ideally, we should have used a CoordinatedShutdown phase for it,
+    // but we can't know if the ActorSystem is going to be shutdown together with this Factory
+    implicit val exc = executionContext
+
+    val stopped =
+      releaseInternalResources()
+        // we don't want to fail the Future if we can't close the internal resources
+        .recover[Either[Throwable, Done]]  {
+            case NonFatal(ex) =>
+              log.warn("failed to close internal resources", ex)
+              Right(Done)
+            case fatal =>
+              log.warn("failed to close internal resources", fatal)
+              Left(fatal)
+        }
+        .flatMap {
+          case Right(_) => actorSystem.terminate()
+          case Left(fatal) =>
+            // if releaseInternalResources throws a fatal exception
+            // we still try to stop the actor system, but fail the final Future
+            // using the fatal exception we got from above
+            actorSystem.terminate()
+              .recover { case NonFatal(_) => throw fatal }
+              .flatMap( _ => throw fatal)
+        }
+        .map(_ => ())
```

Comment:
So, here is the new impl. We close the internal resources and we handle the special case that if a fatal exception is thrown (we never know). 

When the first future completes, we terminate the actor system. If we get a fatal exception. from above we still try to terminate it, but ultimately we fail the future with the fatal exception.

Am I being overzealous?

Reply:
Thanks for reviewing this over and over.

BTW: This idiom is beautiful!